device/esp_tinyusb/test_app/main/test_descriptors_config.c

```diff
@@ -232,4 +232,262 @@ TEST_CASE("descriptors_config_all_configured", "[esp_tinyusb][usb_device]")
+    TUD_CONFIG_DESCRIPTOR(1, 1, 0, TUSB_MSC_CONFIG_DESC_TOTAL_LEN, TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP, 100),
+    TUD_MSC_DESCRIPTOR(1, 0, 0x01, 0x81, 64),
+};
+
+bool tud_msc_test_unit_ready_cb(uint8_t lun)
+{
+    (void) lun;
+    // Our device doesn't have any media for testing purpose
+    return false;
+}
+
+void tud_msc_capacity_cb(uint8_t lun, uint32_t *block_count, uint16_t *block_size)
+{
+    (void) lun;
+    *block_count = 0;
+    *block_size  = 0;
+}
+
+void tud_msc_inquiry_cb(uint8_t lun, uint8_t vendor_id[8], uint8_t product_id[16], uint8_t product_rev[4])
+{
+    (void) lun;
+    const char vid[] = "USB";
+    const char pid[] = "MSC Class";
+    const char rev[] = "0.1";
+
+    memcpy(vendor_id, vid, strlen(vid));
+    memcpy(product_id, pid, strlen(pid));
+    memcpy(product_rev, rev, strlen(rev));
+}
+
+int32_t tud_msc_read10_cb(uint8_t lun, uint32_t lba, uint32_t offset, void *buffer, uint32_t bufsize)
+{
+    return 0;
+}
+
+int32_t tud_msc_write10_cb(uint8_t lun, uint32_t lba, uint32_t offset, uint8_t *buffer, uint32_t bufsize)
+{
+    return 0;
+}
+
+int32_t tud_msc_scsi_cb(uint8_t lun, uint8_t const scsi_cmd[16], void *buffer, uint16_t bufsize)
+{
+    return 0;
+}
+
+//
+// Configuration array
+//
+
+enum {
+    TEST_CFG_HID_CLASS = 0,
+    TEST_CFG_CDC_CLASS,
+    TEST_CFG_MSC_CLASS,
+};
+
+const tinyusb_config_t tusb_cfg[] = {
+    // HID Class
+    {
+        .external_phy = false,
+        .device_descriptor = &test_hid_class_dev_desc,
+        .string_descriptor = test_hid_string_descriptor,
+        .string_descriptor_count = sizeof(test_hid_string_descriptor) / sizeof(test_hid_string_descriptor[0]),
+        .configuration_descriptor = test_hid_fs_configuration_descriptor,
+#if (TUD_OPT_HIGH_SPEED)
+        .hs_configuration_descriptor = test_hid_hs_configuration_descriptor,
+#endif // TUD_OPT_HIGH_SPEED
+    },
+    // CDC Class
+    {
+        .external_phy = false,
+        .device_descriptor = &test_cdc_class_dev_desc,
+        .string_descriptor = test_cdc_string_descriptor,
+        .string_descriptor_count = sizeof(test_cdc_string_descriptor) / sizeof(test_cdc_string_descriptor[0]),
+        .configuration_descriptor = test_cdc_fs_configuration_descriptor,
+#if (TUD_OPT_HIGH_SPEED)
+        .hs_configuration_descriptor = test_cdc_hs_configuration_descriptor,
+#endif // TUD_OPT_HIGH_SPEED
+    },
+    //  MSC Class
+    {
+        .external_phy = false,
+        .device_descriptor = &test_msc_class_dev_desc,
+        .string_descriptor = test_msc_string_descriptor,
+        .string_descriptor_count = sizeof(test_msc_string_descriptor) / sizeof(test_msc_string_descriptor[0]),
+        .configuration_descriptor = test_msc_fs_configuration_descriptor,
+#if (TUD_OPT_HIGH_SPEED)
+        .hs_configuration_descriptor = test_msc_hs_configuration_descriptor,
+#endif // TUD_OPT_HIGH_SPEED
+    },
+};
+
+
+TEST_CASE("device_class_reconfiguration", "[usb_device]")
```

Comment:
@peter-marcisovsky 

After adding the feature from here: https://github.com/espressif/tinyusb/pull/27, we will be able to run `esp_tinyusb` device tests, because they could be run one by one in a one test launch. 

Currently, I have no idea how to add them to our CI and where, but if you have some ideas, please feel free to share. 
Thanks.

Reply:
To add usb device test, you just need to create a new pytest marker in your pytest*.py file, like so:

`@pytest.mark.usb_device`

Then, run pytest from `build_and_run_test_app_usb.yml` file with this new marker. 
Right now, we are using only `-m usb_host` marker. Thus there will be one more marker `-m usb_device`
So the new pytest command would look like:
`pytest --target=${{ matrix.idf_target }} -m usb_host -m usb_device --build-dir=build_${{ matrix.idf_target }}`

This should be all.